### PR TITLE
Feature/1156 mq as uri param for metadata

### PR DESCRIPTION
**Pull request by kzangeli**
Fixes part of issue #1156.
New URI param **mq**.

### CODE REVIEW

File: src/lib/ngsi/Scope.h [hunk 1]
```diff
@@ -44,6 +44,8 @@
 #define SCOPE_FILTER_NOT_EXISTENCE   SCOPE_FILTER "::" SCOPE_OPERATOR_NOT "::" "Existence"
 
 #define SCOPE_TYPE_SIMPLE_QUERY      "FIWARE::StringQuery"
+#define SCOPE_TYPE_SIMPLE_QUERY_MD   "FIWARE::StringQuery::Metadata"
```
Comment (fgalan) [inline src/lib/ngsi/Scope.h]: FIWARE::StringQuery::Metadata should be described in documentation (filters.md?). Probably, the same section used for FIWARE::StringQuery could be used as base (copy-paste-modify).

File: test/functionalTest/cases/1080_batch_query_context/batch_query_context_scopes.test [hunk 2]
```diff
@@ -146,8 +144,8 @@ echo
 echo
 
 
-echo "08. POST /v2/op/query, with E.* with scopes q:!type and q=A5==45, see E5/none"
```
Comment (fgalan) [inline test/functionalTest/cases/1080_batch_query_context/batch_query_context_scopes.test]: An old leftover? Not sure, but I think that `!type` comes from times where enity type optionality was possible, but now that is not possible (given that although the user creates an entity without type CB will set a default type for it).

Is my reasoning correct?

File: CHANGES_NEXT_RELEASE [hunk 3]
```diff
@@ -19,6 +19,7 @@
 - Fix: Allowing trailing semicolon in a string filter, but only one (Issue #2086)
 - Fix: NGSIv2 geoquery syntax errors using 400 Bad Request instead of 422 Unprocessable Entity (according to rules defined in #1286)
 - Fix: NGSIv2 URI PATH must be all in lowercase, otherwise "service not found" (Issue #2057)
+- Add: New URI param 'mq' for matching in metadata (Issue #1156)
```
Comment (fgalan) [inline CHANGES_NEXT_RELEASE]: Actually, what you have achieve in this PR is much larger :), covering not only "mq" in URI param for GET but also FIWARE::StringQuery::Metadata.

Thus, I'd suggest the following 2 lines:

- Add: New param 'mq' (both in URI param for list all entities operation and expression sub-field in subscriptions) for matching in metadata (#1556)
- Add: FIWARE::StringQuery::Metadata scope (#1556)

File: src/lib/jsonParse/jsonQueryContextRequest.cpp [hunk 4]
```diff
@@ -240,7 +240,7 @@ static std::string scopeValue(const std::string& path, const std::string& value,
 
     reqDataP->qcr.scopeP->value = value;
 
-    reqDataP->qcr.scopeP->stringFilterP = new StringFilter();
+    reqDataP->qcr.scopeP->stringFilterP = new StringFilter(SftQ);
```
Comment (fgalan) [inline src/lib/jsonParse/jsonQueryContextRequest.cpp]: I have seen a couple of cases previously:

```
new StringFilter()
```

now

```
new StringFilter(SftQ)
```

Could you elaborate a little bit on this new parameter in StringFilter constrution, please? (maybe at skype)

File: src/lib/rest/StringFilter.cpp [hunk 5]
```diff
@@ -976,32 +1225,59 @@ bool StringFilter::mongoFilterPopulate(std::string* errorStringP)
...
+      }
+
+      if (itemP->metadataName == "")
+      {
+        *errorStringP = "no metadata name - not valid for metadata filters";
+        return false;
+      }
 
-    if (itemP->left == DATE_CREATED)
+      left = itemP->attributeName + ".md." + itemP->metadataName;
```
Comment (fgalan) [inline src/lib/rest/StringFilter.cpp]: `md` is due to the name it takes from DB. Thus, don't use it literralry, but CSUB_EXPR_MQ.

File: src/lib/rest/StringFilter.cpp [hunk 6]
```diff
@@ -1307,7 +1671,18 @@ bool StringFilter::match(ContextElementResponse* cerP)
       break;
 
     case SfopMatchPattern:
-      return itemP->matchPattern(caP);
+      //
+      // To not be forgotten for CNR (and discussed during PR-review):
```
Comment (fgalan) [inline src/lib/rest/StringFilter.cpp]: By your comprehensive description, I agree this is a bugfix and should be mentioned in the CNR (and probably covered by a .test).

(Side-note: it seem this bug was introduced by me when I implemented `~=` filter (https://github.com/telefonicaid/fiware-orion/blame/develop/src/lib/rest/StringFilter.cpp#L1310). I didn't realized of the "return pattern" in above cases and trying to achieve something more compact I introduced the bug).

File: src/lib/rest/StringFilter.cpp [hunk 7]
```diff
@@ -1323,7 +1698,7 @@ bool StringFilter::match(ContextElementResponse* cerP)
 */
 StringFilter* StringFilter::clone(std::string* errorStringP)
 {
-  StringFilter* sfP = new StringFilter();
+  StringFilter* sfP = new StringFilter(type);
```
Comment (fgalan) [inline src/lib/rest/StringFilter.cpp]: Another case of `new StringFilter()` that has become in `new StringFilter(something)`. I wonder why...

File: src/lib/rest/rest.cpp [hunk 8]
```diff
@@ -240,7 +240,7 @@ static int uriArgumentGet(void* cbDataP, MHD_ValueKind kind, const char* ckey, c
   {
     containsForbiddenChars = forbiddenChars(val, ";");
   }
-  else if ((key != "q") && (key != "idPattern"))
+  else if ((key != "q") && (key != "mq") && (key != "idPattern"))
```
Comment (fgalan) [inline src/lib/rest/rest.cpp]: In line 210 you are using URI_PARAM_Q and URI_PARAM_MQ but here you are using "q" and "mq"... Is there a reason for that? Or just a pending refactor to align existing "q", "mq", etc. to their respective URI_PARAM_* ?

File: src/lib/serviceRoutinesV2/getEntities.cpp [hunk 9]
```diff
@@ -82,6 +83,7 @@ std::string getEntities
   std::string  id         = ciP->uriParam["id"];
   std::string  idPattern  = ciP->uriParam["idPattern"];
   std::string  q          = ciP->uriParam["q"];
+  std::string  mq         = ciP->uriParam["mq"];
```
Comment (fgalan) [inline src/lib/serviceRoutinesV2/getEntities.cpp]: Again, if "mq" refers to the name of an URI param (it seems so by `uriParam` :)) the URI_PARAM_MQ should be used. Is that correct?

File: test/functionalTest/cases/1156_mq_as_uri_param_for_metadata/mq_as_uri_param_for_metadata.test [hunk 10]
```diff
@@ -0,0 +1,807 @@
...
+
+#
+# 01. Create an entity E1 with attr A1 with metadata M1=12
+# 02. Create an entity E2 with attr A1 with metadata M1=13
+# 03. Create an entity E3 with attr A1 with metadata M1=14
+# 04. Lookup all entities having A1.M1==12, see E1
+# 05. Create a subscription for all entities/A1 AND mq=A1.M1==13
+# 06. Dump accumulator, see E2, then reset accumulator
+# 07. Modify E1/A1, setting A1/M1=13
+# 08. Dump accelerator, see E1, then reset accumulator
```
Comment (fgalan) [inline test/functionalTest/cases/1156_mq_as_uri_param_for_metadata/mq_as_uri_param_for_metadata.test]: accelerator -> accumulator

Check also corresponding test in echoes and expected output. Also in step 10.

File: test/functionalTest/cases/1156_mq_as_uri_param_for_metadata/mq_as_uri_param_for_metadata.test [hunk 11]
```diff
@@ -0,0 +1,807 @@
...
+accumulatorStart --pretty-print
+
+--SHELL--
+
+#
+# 01. Create an entity E1 with attr A1 with metadata M1=12
+# 02. Create an entity E2 with attr A1 with metadata M1=13
+# 03. Create an entity E3 with attr A1 with metadata M1=14
+# 04. Lookup all entities having A1.M1==12, see E1
+# 05. Create a subscription for all entities/A1 AND mq=A1.M1==13
```
Comment (fgalan) [inline test/functionalTest/cases/1156_mq_as_uri_param_for_metadata/mq_as_uri_param_for_metadata.test]: I'd suggest to add a GET /v2/subscriptions/subId step, to check that mq is correctly rendered.

File: test/functionalTest/cases/1156_mq_as_uri_param_for_metadata/mq_as_uri_param_for_metadata.test [hunk 12]
```diff
@@ -0,0 +1,807 @@
...
+# iot_support at tid dot es
+
+# VALGRIND_READY - to mark the test ready for valgrindTestSuite.sh
+
+--NAME--
+mq as URI param for metadata in filters
+
+--SHELL-INIT--
+dbInit CB
+brokerStart CB
```
Comment (fgalan) [inline test/functionalTest/cases/1156_mq_as_uri_param_for_metadata/mq_as_uri_param_for_metadata.test]: Maybe you have already done but just in case... have you tested this .test with brokerStart adjusted for -noCache? Just to be sure that it also works in that way.

(I know, manually testing the -noCache cases is not the best approach... but the only possible right now)

File: src/lib/jsonParse/jsonQueryContextRequest.cpp (hunk 4 above)
Comment (kzangeli) [inline src/lib/jsonParse/jsonQueryContextRequest.cpp]: Explained over skype

NTC

File: test/functionalTest/cases/1156_mq_as_uri_param_for_metadata/q_and_mq_as_uri_param_for_metadata.test [hunk 13]
```diff
@@ -0,0 +1,454 @@
...
+brokerStart CB
+accumulatorStart --pretty-print
+
+--SHELL--
+
+#
+# 01. Create entity E1, with attr A1=1, having metadata M1=1
+# 02. Create entity E2, with attr A1=1, having metadata M1=2
+# 03. Create entity E3, with attr A1=1, having metadata M2=3
+# 04. GET entities matching q=A1==1 AND mq=A1.M1==2, see E2
```
Comment (fgalan) [inline test/functionalTest/cases/1156_mq_as_uri_param_for_metadata/q_and_mq_as_uri_param_for_metadata.test]: The `q` uses in all the cases will always match all entities. Maybe you should create some additional entities with A1 different from 1 and with A2 attribute instead of A1 in to check that the test is not capturing them.

File: test/functionalTest/cases/1156_mq_as_uri_param_for_metadata/q_and_mq_as_uri_param_for_metadata.test [hunk 14]
```diff
@@ -0,0 +1,454 @@
...
+    ], 
+    "subscriptionId": "REGEX([0-9a-f]{24})"
+}
+=======================================
+
+
+--TEARDOWN--
+brokerStop CB
+dbDrop CB
+accumulatorStop
```
Comment (fgalan) [inline test/functionalTest/cases/1156_mq_as_uri_param_for_metadata/q_and_mq_as_uri_param_for_metadata.test]: I'd suggest to add a couple more .test:

* Test "FIWARE:StringQuery::Metadata" in /v1/queryContext scope
* Test "FIWARE::StringQuery::Metadata" in /v2/op/query scope

File: test/unittests/mongoBackend/mongoQueryContext_filters_test.cpp [hunk 15]
```diff
@@ -1225,6 +1228,7 @@ TEST(mongoQueryContextRequest_filters, cobingingSeveralFilters)
 
   /* Invoke the function in mongoBackend library */
   servicePathVector.clear();
+  // servicePathVector.push_back("/#");  Line to be removed during PR
```
Comment (fgalan) [inline test/unittests/mongoBackend/mongoQueryContext_filters_test.cpp]: As discussed at skype, remove all lines of this kind.

File: src/lib/rest/rest.cpp (hunk 8 above)
Comment (kzangeli) [inline src/lib/rest/rest.cpp]: Bah, it was 'q' before, I just added mq ...
Time to fix them, and "idPattern" also.

File: test/functionalTest/cases/1080_batch_query_context/batch_query_context_scopes.test (hunk 2 above)
Comment (kzangeli) [inline test/functionalTest/cases/1080_batch_query_context/batch_query_context_scopes.test]: This was simply a mistake in the description of the step. The filter is actually about 'existance of B', it was changed when enity type optionality was revoked ...
Now it is fixed.

NTC

File: test/functionalTest/cases/1156_mq_as_uri_param_for_metadata/mq_as_uri_param_for_metadata.test (hunk 12 above)
Comment (kzangeli) [inline test/functionalTest/cases/1156_mq_as_uri_param_for_metadata/mq_as_uri_param_for_metadata.test]: It would not be that difficult to modify the testHarness to accomplish this ...
A new parameter `--noCache`, making the testHarness always add  ' --noCache`, when starting the broker.
We could have jenkins run the functests twice., Once with sub-cache and once without. We'd have to remove all 'already covered cases' first, of course. With 'covered cases' I mean functests that already exist in both versions: with and without sub-cache activated.

That said, no, I haven't started the tests without sub.cache.
But I will do it :-)

File: test/functionalTest/cases/1156_mq_as_uri_param_for_metadata/mq_as_uri_param_for_metadata.test (hunk 10 above)
Comment (kzangeli) [inline test/functionalTest/cases/1156_mq_as_uri_param_for_metadata/mq_as_uri_param_for_metadata.test]: accelerator ... :-)
I write accelerator instead of accumulator with a frequency of around 25% :-D

File: src/lib/ngsi/Scope.h (hunk 1 above)
Comment (kzangeli) [inline src/lib/ngsi/Scope.h]: Fixed in a8f0e82

File: CHANGES_NEXT_RELEASE (hunk 3 above)
Comment (kzangeli) [inline CHANGES_NEXT_RELEASE]: Fixed in a8f0e82

File: src/lib/rest/StringFilter.cpp (hunk 5 above)
Comment (kzangeli) [inline src/lib/rest/StringFilter.cpp]: Sure, however, CSUB_EXPR_MQ is not the right constant.
We need a constant for name of the vector of the metadata used in mongo. 
I guess ENT_ATTRS_MD is what should be used ...

File: src/lib/rest/StringFilter.cpp (hunk 6 above)
Comment (kzangeli) [inline src/lib/rest/StringFilter.cpp]: Yeah. I added the NOTE in the function header, to try to minimize the risk of this happening again.
Probably nobody will read it ... :-D

File: src/lib/rest/StringFilter.cpp (hunk 7 above)
Comment (kzangeli) [inline src/lib/rest/StringFilter.cpp]: Discussed over skype.
However, I can mention it here also ...

As there are now two flavors of StringFilter ('q' and 'mq'), when creating an instance of the class, the instance must be initialized to indicate the flavor. 

NTC

File: test/unittests/mongoBackend/mongoQueryContext_filters_test.cpp (hunk 15 above)
Comment (kzangeli) [inline test/unittests/mongoBackend/mongoQueryContext_filters_test.cpp]: Fixed in bad5764

File: test/functionalTest/cases/1156_mq_as_uri_param_for_metadata/mq_as_uri_param_for_metadata.test (hunk 10 above)
Comment (kzangeli) [inline test/functionalTest/cases/1156_mq_as_uri_param_for_metadata/mq_as_uri_param_for_metadata.test]: Fixed in ee0fa7f  (including an older functest - surprised I found only *one* more ... :-))

File: test/functionalTest/cases/1156_mq_as_uri_param_for_metadata/mq_as_uri_param_for_metadata.test (hunk 11 above)
Comment (kzangeli) [inline test/functionalTest/cases/1156_mq_as_uri_param_for_metadata/mq_as_uri_param_for_metadata.test]: Not a bad idea, indeed :-)
The rendering of mq was not implemented.
Fixed in 68ea322

File: src/lib/serviceRoutinesV2/getEntities.cpp (hunk 9 above)
Comment (kzangeli) [inline src/lib/serviceRoutinesV2/getEntities.cpp]: That is correct.
Fixed in f3b2e4a

File: src/lib/rest/rest.cpp (hunk 8 above)
Comment (kzangeli) [inline src/lib/rest/rest.cpp]: Fixed in f3b2e4a

Only q and mq - to not mix things too much in this PR.
There are plenty of other URI params (like idPattern) that use simple strings. Most of them don't even have a define in uriParamNames.h. This 'problem' should probably be addressed in another PR.

File: src/lib/rest/StringFilter.cpp (hunk 5 above)
Comment (kzangeli) [inline src/lib/rest/StringFilter.cpp]: Fixed in f3b2e4a

File: src/lib/rest/StringFilter.cpp (hunk 6 above)
Comment (kzangeli) [inline src/lib/rest/StringFilter.cpp]: Functest implemented in 258e272

File: test/functionalTest/cases/1156_mq_as_uri_param_for_metadata/q_and_mq_as_uri_param_for_metadata.test (hunk 14 above)
Comment (kzangeli) [inline test/functionalTest/cases/1156_mq_as_uri_param_for_metadata/q_and_mq_as_uri_param_for_metadata.test]: Fixed in 65f8be1

File: test/functionalTest/cases/1156_mq_as_uri_param_for_metadata/q_and_mq_as_uri_param_for_metadata.test (hunk 13 above)
Comment (kzangeli) [inline test/functionalTest/cases/1156_mq_as_uri_param_for_metadata/q_and_mq_as_uri_param_for_metadata.test]: Fixed in 65f8be1

File: test/functionalTest/cases/1156_mq_as_uri_param_for_metadata/mq_as_uri_param_for_metadata.test (hunk 12 above)
Comment (kzangeli) [inline test/functionalTest/cases/1156_mq_as_uri_param_for_metadata/mq_as_uri_param_for_metadata.test]: Done, and a bugfix added in 65f8be1

File: CHANGES_NEXT_RELEASE [hunk 16]
```diff
@@ -19,6 +19,10 @@
 - Fix: Allowing trailing semicolon in a string filter, but only one (Issue #2086)
 - Fix: NGSIv2 geoquery syntax errors using 400 Bad Request instead of 422 Unprocessable Entity (according to rules defined in #1286)
 - Fix: NGSIv2 URI PATH must be all in lowercase, otherwise "service not found" (Issue #2057)
+- Fix: A bug fixed in URI param 'q', avoiding false matches when using string match operator (~=) that matches AND some operator that does not match.
+- Add: New URI param 'mq' for matching in metadata (Issue #1156) 
+- Add: New param 'mq' (both in URI param for list all entities operation and expression sub-field in subscriptions) for matching in metadata (#1156)
```
Comment (fgalan) [inline CHANGES_NEXT_RELEASE]: Line 23 is redundant, already covered by lines 24. Suggestion: just remove line 23.

File: CHANGES_NEXT_RELEASE [hunk 17]
```diff
@@ -39,6 +43,7 @@
 - Fix: Correct error returned when trying to replace a non-existing attribute in PUT /v2/entities/<entity-id>/attrs/<attr-name> (Issue #2221)
 - Fix: Empty string detection for entity id, entity type and attribute name in NGSIv2 request URLs (#1426, #1487)
 - Fix: Trailing slashes in URI path no longer silently removed, but considered and giving adequate error responses
+- Fix: Initial notifications were counted twice (which gave erroneous timesSent in GET /v2/subscriptions/<subId>)
```
Comment (fgalan) [inline CHANGES_NEXT_RELEASE]: Under which circustances? Only in the case of using case? In that case, the line should be adjusted:

Fix: Initial notifications were counted twice when csub cache was used (which gave erroneous timesSent in GET /v2/subscriptions/<subId>)

File: src/lib/rest/StringFilter.cpp (hunk 5 above)
Comment (fgalan) [inline src/lib/rest/StringFilter.cpp]: Yes, that's correct. Right one is ENT_ATTRS_MD

File: src/lib/rest/StringFilter.cpp [hunk 18]
```diff
@@ -1307,7 +1673,18 @@ bool StringFilter::match(ContextElementResponse* cerP)
...
 
     case SfopMatchPattern:
-      return itemP->matchPattern(caP);
+      //
+      // To not be forgotten for CNR (and discussed during PR-review):
+      //   This here is a bug fix.
+      //   If a 'q/mq' contains two expressions, a 'pattern' and an 'eq', and the eq gives false,
+      //   while the pattern gives true, then the total outcome will be true (before the bug fix),
+      //   because of the inconditional return that the bugfix removes.
+      //   The items in 'q/mq' are ANDed and as true AND false gives false, this was a bug.
```
Comment (fgalan) [inline src/lib/rest/StringFilter.cpp]: Now that we have discussed the above and added the entry to CNR I understand that the comment could be removed

File: src/lib/rest/rest.cpp (hunk 8 above)
Comment (fgalan) [inline src/lib/rest/rest.cpp]: I agree. I'd suggest to create a new issue (techdebt nice2have about this) so we don't forgot it in the future.

File: src/lib/rest/StringFilter.cpp (hunk 18 above)
Comment (kzangeli) [inline src/lib/rest/StringFilter.cpp]: Fixed in 9f2b135

File: CHANGES_NEXT_RELEASE (hunk 17 above)
Comment (kzangeli) [inline CHANGES_NEXT_RELEASE]: Fixed in 9f2b135

File: CHANGES_NEXT_RELEASE (hunk 16 above)
Comment (kzangeli) [inline CHANGES_NEXT_RELEASE]: Fixed in 9f2b135

File: doc/manuals/user/filtering.md [hunk 19]
```diff
@@ -188,7 +188,26 @@ described in detail in [the following section](geolocation.md#geo-located-querie
 
 [Top](#top)
 
-### String filter
+### String filters
```
Comment (fgalan) [inline doc/manuals/user/filtering.md]: As a consecuence of changing the tittle references to `#string-filter` has to change to `#string-filters`. I have located at least two ones:

```
fermin@neodeb:~/src/fiware-orion/doc/manuals$ grep string-filter -R *
user/v1_v2_coexistence.md:[the following section](filtering.md#string-filter) for details.
user/filtering.md:  * [String query filter](#string-filter)
```

File: doc/manuals/user/filtering.md [hunk 20]
```diff
@@ -188,7 +188,26 @@ described in detail in [the following section](geolocation.md#geo-located-querie
...
+
+Two different types of string filters are supported:
+* filtering over attributes (scope type: FIWARE::StringQuery) and
```
Comment (fgalan) [inline doc/manuals/user/filtering.md]: Looking to the style in other places of this file, filter types are surrounded by quotes. See line 186, for instance.

(Also to fix in the next line)

File: doc/manuals/user/filtering.md [hunk 21]
```diff
@@ -188,7 +188,26 @@ described in detail in [the following section](geolocation.md#geo-located-querie
...
+* filtering over the metadatas of attributes (scope type: FIWARE::StringQuery::Metadata).
+
+These scopes allow to express filtering conditions such as equality, unequality,
+greater/less than, in-array, range or existence (of attributes and metadata of attributes).
+
+There isn't any URL parameter corresponding to these filters in NGSIv1.
+In NGSIv2 they correspond to the `q` and `mq` URI parameters.
+
+For a detailed syntax description of the `value` or `q` and `mq` parameters, see [NGSIv2 specification
```
Comment (fgalan) [inline doc/manuals/user/filtering.md]: "value" should be in normal text (or I'm missing the reason to use code-style in this case...)

File: doc/manuals/user/forbidden_characters.md [hunk 22]
```diff
@@ -51,6 +51,7 @@ are involved, anyway.
 There are some exception cases in which the above restrictions do not apply. In particular, in the following fields:
 
 * URL parameter `q` and the value of "FIWARE::StringQuery" scope allow the special characters needed by the Simple Query Language
+* URL parameter `mq` and the value of "FIWARE::StringQuery::Metadata" scope allow the special characters needed by the Simple Query Language
```
Comment (fgalan) [inline doc/manuals/user/forbidden_characters.md]: I didn't remember this piece of documentation... Good catch! :)

NTC

File: doc/manuals/user/filtering.md (hunk 21 above)
Comment (kzangeli) [inline doc/manuals/user/filtering.md]: It was like that in the initial text (that I moved or copied).
I agree that is looks a bit strange. 

Perhaps: For a detailed syntax description of the value of the string filters, and the `q` and `mq` parameters, see ...

File: doc/manuals/user/filtering.md (hunk 19 above)
Comment (kzangeli) [inline doc/manuals/user/filtering.md]: Fixed in 187d5e2

File: doc/manuals/user/filtering.md (hunk 20 above)
Comment (kzangeli) [inline doc/manuals/user/filtering.md]: Fixed in 187d5e2

File: doc/manuals/user/filtering.md (hunk 21 above)
Comment (kzangeli) [inline doc/manuals/user/filtering.md]: Fixed in 187d5e2

**Comment by fgalan** [comment]
LGTM
